dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +210,254 @@ cdef int _local_tracker(
+        dir[j] = first_step[j]
+        vs[j] = voxel_size[j]
+    copypoint(dir, &directions[0, 0])
+
+    tissue_class[0] = TRACKPOINT
+    i = 0
+    while i < streamline.shape[0] - 1:
+        if dg.get_direction(point, dir):
+            # no valid diffusion directions to follow
+            tissue_class[0] = INVALIDPOINT
+        else:
+            for j in range(3):
+                voxdir[j] = dir[j] / vs[j]
+            i += 1
+            fixed_step(point, voxdir, step_size)
+            copypoint(point, & streamline[i, 0])
+            copypoint(dir, & directions[i, 0])
+
+            tissue_class[0] = tc.check_point_c(point)
+
+        if tissue_class[0] == TRACKPOINT:
+            continue
+        elif tissue_class[0] == ENDPOINT:
+            i += 1
+            break
+        elif tissue_class[0] == INVALIDPOINT:
+            if pft_trial < pft_max_trial and i > 1:
+                pft_streamline_i = min(i - 1, pft_nbr_back_steps)
+                pft_nbr_steps = min(pft_max_steps,
+                                    streamline.shape[0] - pft_streamline_i - 1)
+                i = _pft(streamline,
+                         pft_streamline_i,
+                         directions,
+                         dg,
+                         tc,
+                         voxel_size,
+                         step_size,
+                         tissue_class,
+                         pft_nbr_steps,
+                         particle_count,
+                         particle_paths,
+                         particle_dirs,
+                         particle_weights,
+                         particle_states)
+
+                pft_trial += 1
+                # update the current point with the PFT results
+                for j in range(3):
+                    point[j] = streamline[i, j]
+                    dir[j] = directions[i, j]
+
+                if not tissue_class[0] == TRACKPOINT:
+                    break
+            else:
+                i += 1
+                break
+        elif tissue_class[0] == OUTSIDEIMAGE:
+            break
+    return i
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef _pft(np.float_t[:,:]  streamline,
+          int streamline_i,
+          np.float_t[:,:]  directions,
+          DirectionGetter dg,
+          ConstrainedTissueClassifier tc,
+          np.float_t[:]  voxel_size,
+          double step_size,
+          TissueClass* tissue_class,
+          int pft_nbr_steps,
+          int particle_count,
+          np.float_t[:,:,:,:] particle_paths,
+          np.float_t[:,:,:,:] particle_dirs,
+          np.float_t[:,:] particle_weights,
+          np.int_t[:,:,:]  particle_states):
+    cdef:
+        double sum_weights, sum_squared, N_effective, rdm_sample#
+        double point[3], dir[3], vs[3], voxdir[3]
+        int s, p, j
+
+    if pft_nbr_steps <= 0:
+        return INVALIDPOINT, streamline_i
+
+    for j in range(3):
+        vs[j] = voxel_size[j]
```

Comment:
It's done everywhere like this but... isn't this a `copypoint`?

Reply:
@jchoude had the same comment below.

here`voxel_size` is of type float_t[:], while copypoint(.,.) needs double *. Is there a more elegant way to do it?

I'm changing all the `for j in range(3):` everywhere with types `double*`. Thx.

Reply:
Thank you for the `copy_point` modif, it's much better. For `voxel_size`, it would be better if it was converted to float64 one time in the `main()`. Again, it's probably negligible, but it's a waste anyway.